ext/serializer.c

```diff
@@ -1256,6 +1256,25 @@ void ddtrace_serialize_span_to_array(ddtrace_span_data *span, zval *array) {
         }
 
         add_assoc_zval(el, "service", &prop_service_as_string);
+
+        // Add _dd.base_service if service name differs from mapped root service name
+        zval *prop_root_service = &span->root->property_service;
+        ZVAL_DEREF(prop_root_service);
+        zval prop_root_service_as_string;
+        ddtrace_convert_to_string(&prop_root_service_as_string, prop_root_service);
+
+        zval *new_root_name = zend_hash_find(service_mappings, Z_STR(prop_root_service_as_string));
+        if (new_root_name) {
+            zend_string_release(Z_STR(prop_root_service_as_string));
+            ZVAL_COPY(&prop_root_service_as_string, new_root_name);
+        }
+
+        if (strcasecmp(Z_STRVAL(prop_service_as_string), Z_STRVAL(prop_root_service_as_string)) != 0) {
+            zend_array *meta = ddtrace_property_array(&span->property_meta);
+            zend_hash_str_add_new(meta, ZEND_STRL("_dd.base_service"), &prop_root_service_as_string);
```

Comment:
```suggestion
            zend_hash_str_update(meta, ZEND_STRL("_dd.base_service"), &prop_root_service_as_string);
```
Don't use `_add_new` if you cannot be sure that there's no user-input you could overwrite. If there's already a manually added meta key from PHP code named _dd.base_service, it'll leak. And just `_add` needs to be checked for success or failure to potentially free the value.